aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,607 @@
+              iheight,
+              owidth,
+              oheight);
+        }));
+  }
+  return grad_input;
+}
+
+Tensor max_unpooling2d_backward_cpu(
+    const Tensor& grad_output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  auto grad_input = at::empty_like(self);
+  max_unpooling2d_backward_out_cpu(
+      grad_input, grad_output, self, indices, output_size);
+  return grad_input;
+}
+
+template <typename scalar_t>
+static void max_unpooling3d_backward_out_cpu_frame(
+    scalar_t* gradInput_p,
+    scalar_t* gradOutput_p,
+    int64_t* ind_p,
+    int64_t nslices,
+    int64_t iT,
+    int64_t iW,
+    int64_t iH,
+    int64_t oT,
+    int64_t oW,
+    int64_t oH) {
+  int k;
+  bool has_error = false;
+  int error_index = 0;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++) {
+    scalar_t* gradInput_p_k = gradInput_p + k * iT * iH * iW;
+    scalar_t* gradOutput_p_k = gradOutput_p + k * oT * oH * oW;
+    int64_t* ind_p_k = ind_p + k * iT * iH * iW;
+
+    int t, i, j, index;
+    int64_t maxp;
+    for (t = 0; t < iT; t++) {
+      for (i = 0; i < iH; i++) {
+        for (j = 0; j < iW; j++) {
+          index = t * iH * iW + i * iW + j;
+          maxp = ind_p_k[index]; /* retrieve position of max */
+          if (maxp < 0 || maxp >= oT * oH * oW) {
+#pragma omp critical
+            {
+              has_error = true;
+              error_index = maxp;
+            }
+          }
+          gradInput_p_k[index] = gradOutput_p_k[maxp]; /* update gradient */
+        }
+      }
+    }
+  }
+  if (has_error) {
+    AT_ERROR(
+        "invalid max index ",
+        error_index,
+        ", oT= ",
+        oT,
+        ", oW= ",
+        oW,
+        ",oH= ",
+        oH);
+  }
+}
+
+Tensor& max_unpooling3d_backward_out_cpu(
+    Tensor& grad_input,
+    const Tensor& grad_output_,
+    const Tensor& self,
+    const Tensor& indices_,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  auto oT = output_size[0];
+  auto oH = output_size[1];
+  auto oW = output_size[2];
+  int dimw = 3;
+  int dimh = 2;
+  int dimt = 1;
+  int nbatch = 1;
+  int nslices;
+  int iT;
+  int iH;
+  int iW;
+
+  max_unpooling3d_shape_check(
+      self, grad_output_, indices_, output_size, stride, padding);
+
+  // TODO (from THNN): check gradOutput shape
+  /* get contiguous gradOutput */
+  auto grad_output = grad_output_.contiguous();
+  auto indices = indices_.contiguous();
+
+  /* resize */
+  grad_input.resize_as_(self);
+  grad_input.zero_();
+  if (self.ndimension() == 5) {
+    nbatch = self.size(0);
+    dimt++;
+    dimw++;
+    dimh++;
+  }
+
+  /* sizes */
+  nslices = self.size(dimt - 1);
+  iT = self.size(dimt);
+  iH = self.size(dimh);
+  iW = self.size(dimw);
+
+  /* backprop */
+  int p;
+  for (p = 0; p < nbatch; p++) {
+    int inputOffset = p * nslices * iT * iH * iW;
+    int outputOffset = p * nslices * oT * oH * oW;
+    AT_DISPATCH_FLOATING_TYPES(
```

Comment:
You're doing a dynamic dispatch for batch in the tensor. Better to the dynamic dispatch once before the batch loop.